# Building a Fake News classifier with Naive Bayes

The idea of this project is to try to build a fake news classifier with Naive Bayes algorithm. I'm using a fake and real news dataset by Clément Bisallion that has 23481 fake news and 21417 real news. The files can be found [here](https://www.kaggle.com/clmentbisaillon/fake-and-real-news-dataset?select=True.csv)

In [1]:
import re
import pandas as pd
pd.options.display.max_colwidth=100

fake = pd.read_csv('Fake.csv')
real = pd.read_csv('True.csv')

In [2]:
fake.head()

title  \
0              Donald Trump Sends Out Embarrassing New Year’s Eve Message; This is Disturbing   
1                        Drunk Bragging Trump Staffer Started Russian Collusion Investigation   
2   Sheriff David Clarke Becomes An Internet Joke For Threatening To Poke People ‘In The Eye’   
3               Trump Is So Obsessed He Even Has Obama’s Name Coded Into His Website (IMAGES)   
4                       Pope Francis Just Called Out Donald Trump During His Christmas Speech   

                                                                                                  text  \
0  Donald Trump just couldn t wish all Americans a Happy New Year and leave it at that. Instead, he...   
1  House Intelligence Committee Chairman Devin Nunes is going to have a bad day. He s been under th...   
2  On Friday, it was revealed that former Milwaukee Sheriff David Clarke, who was being considered ...   
3  On Christmas day, Donald Trump announced that he would  be back to work  the following day, but ...   
4  Pope Francis used his annual Christmas Day message to rebuke Donald Trump without even mentionin...   

  subject               date  
0    News  December 31, 2017  
1    News  December 31, 2017  
2    News  December 30, 2017  
3    News  December 29, 2017  
4    News  December 25, 2017

In [3]:
real.head()

title  \
0       As U.S. budget fight looms, Republicans flip their fiscal script   
1       U.S. military to accept transgender recruits on Monday: Pentagon   
2           Senior U.S. Republican senator: 'Let Mr. Mueller do his job'   
3            FBI Russia probe helped by Australian diplomat tip-off: NYT   
4  Trump wants Postal Service to charge 'much more' for Amazon shipments   

                                                                                                  text  \
0  WASHINGTON (Reuters) - The head of a conservative Republican faction in the U.S. Congress, who v...   
1  WASHINGTON (Reuters) - Transgender people will be allowed for the first time to enlist in the U....   
2  WASHINGTON (Reuters) - The special counsel investigation of links between Russia and President T...   
3  WASHINGTON (Reuters) - Trump campaign adviser George Papadopoulos told an Australian diplomat in...   
4  SEATTLE/WASHINGTON (Reuters) - President Donald Trump called on the U.S. Postal Service on Frida...   

        subject                date  
0  politicsNews  December 31, 2017   
1  politicsNews  December 29, 2017   
2  politicsNews  December 31, 2017   
3  politicsNews  December 30, 2017   
4  politicsNews  December 29, 2017

At first look one of the things people should look to try to identify if a news is fake or real is the first line of text, most of the news starts with the location and media were it was written. I'm going to create a column in both dataframes with the labels fake and real, then i'm going to concatenate both dfs and find the percentage of fake and real news

In [4]:
fake['label'] = 'Fake'
real['label'] = 'Real'

full = pd.concat([fake, real])
full.drop(columns=['text', 'subject', 'date'], inplace=True)
full.head()

title  \
0              Donald Trump Sends Out Embarrassing New Year’s Eve Message; This is Disturbing   
1                        Drunk Bragging Trump Staffer Started Russian Collusion Investigation   
2   Sheriff David Clarke Becomes An Internet Joke For Threatening To Poke People ‘In The Eye’   
3               Trump Is So Obsessed He Even Has Obama’s Name Coded Into His Website (IMAGES)   
4                       Pope Francis Just Called Out Donald Trump During His Christmas Speech   

  label  
0  Fake  
1  Fake  
2  Fake  
3  Fake  
4  Fake

However this set is really big, so the let's take 5% of it

In [5]:
data_randomized = full.sample(frac=1, random_state=1)
full_set_index = round(len(data_randomized) * 0.90)
full_set = data_randomized[full_set_index:].reset_index(drop=True)
full_set.shape[0]

4490

In [6]:
full_set['label'].value_counts(normalize=True)

Fake    0.512695
Real    0.487305
Name: label, dtype: float64

There's a 50/50 distribution between fake and real news. To make the testing process more realistic, i'm keeping 80% of the set for training and 20% for testing.

In [7]:
data_randomized = full_set.sample(frac=1, random_state=1)

training_test_index = round(len(data_randomized) * 0.8)

train = data_randomized[:training_test_index].reset_index(drop=True)
test = data_randomized[training_test_index:].reset_index(drop=True)

print(train.shape[0])
print(test.shape[0])

3592
898


In [8]:
train['label'].value_counts(normalize=True)

Fake    0.504454
Real    0.495546
Name: label, dtype: float64

In [9]:
test['label'].value_counts(normalize=True)

Fake    0.545657
Real    0.454343
Name: label, dtype: float64

After sampling both datasets ended with the same proportion (52/48) of fake/real news

Now let's clean the dataset to train the algorithm

## Cleaning letter, case and punctuation for titles only

In [10]:
#Before cleaning
train.head()

title  \
0                     Trump says to approve lifting restrictions on South Korea missile payload limits   
1  VANISHED: FBI FILES Related To Mysterious “Suicide” Death Of Hillary’s Trusted WH Counsel, Vince...   
2                           WAR ON WORDS: Facebook Censorship Widens, Website to Curate ‘Favored’ News   
3                                     Mexico to review need for tax changes after U.S. reform-document   
4                Harry Reid Has Brutally Honest Message For The GOP From Their Pals At The NRA (TWEET)   

  label  
0  Real  
1  Fake  
2  Fake  
3  Real  
4  Fake

In [11]:
#After cleaning
train['title'] = train['title'].str.replace('\W', ' ')
train['title'] = train['title'].str.lower()
train.head()

title  \
0                     trump says to approve lifting restrictions on south korea missile payload limits   
1  vanished  fbi files related to mysterious  suicide  death of hillary s trusted wh counsel  vince...   
2                           war on words  facebook censorship widens  website to curate  favored  news   
3                                     mexico to review need for tax changes after u s  reform document   
4                harry reid has brutally honest message for the gop from their pals at the nra  tweet    

  label  
0  Real  
1  Fake  
2  Fake  
3  Real  
4  Fake

## Creating vocabulary for title only

In [12]:
train['title'] = train['title'].str.split()
vocabulary = []

for row in train['title']:
    for word in row:
        vocabulary.append(word)
        
vocabulary = list(set(vocabulary))

len(vocabulary)

7716

There are 5431 unique words in the train set. Now let's create a dictionary to count the words in the fake and real news titles

In [13]:
word_counts_per_title = {unique_word: [0] * len(train['title']) for unique_word in vocabulary}

for index, title in enumerate(train['title']):
    for word in title:
        word_counts_per_title[word][index] += 1
        
word_counts_title = pd.DataFrame(word_counts_per_title)
word_counts_title.head()

screams  scotland  epa  baton  enigma  alleviate  unyielding  fulfilling  \
0        0         0    0      0       0          0           0           0   
1        0         0    0      0       0          0           0           0   
2        0         0    0      0       0          0           0           0   
3        0         0    0      0       0          0           0           0   
4        0         0    0      0       0          0           0           0   

   boardrooms  columbia  ...  driven  arabia  adios  arbitrary  cake  drills  \
0           0         0  ...       0       0      0          0     0       0   
1           0         0  ...       0       0      0          0     0       0   
2           0         0  ...       0       0      0          0     0       0   
3           0         0  ...       0       0      0          0     0       0   
4           0         0  ...       0       0      0          0     0       0   

   handler  hills  donation  toilet  
0        0      0         0       0  
1        0      0         0       0  
2        0      0         0       0  
3        0      0         0       0  
4        0      0         0       0  

[5 rows x 7716 columns]

In [14]:
train_clean = pd.concat([train, word_counts_title], axis=1)
train_clean.head()

title  \
0        [trump, says, to, approve, lifting, restrictions, on, south, korea, missile, payload, limits]   
1  [vanished, fbi, files, related, to, mysterious, suicide, death, of, hillary, s, trusted, wh, cou...   
2                   [war, on, words, facebook, censorship, widens, website, to, curate, favored, news]   
3                         [mexico, to, review, need, for, tax, changes, after, u, s, reform, document]   
4  [harry, reid, has, brutally, honest, message, for, the, gop, from, their, pals, at, the, nra, tw...   

  label  screams  scotland  epa  baton  enigma  alleviate  unyielding  \
0  Real        0         0    0      0       0          0           0   
1  Fake        0         0    0      0       0          0           0   
2  Fake        0         0    0      0       0          0           0   
3  Real        0         0    0      0       0          0           0   
4  Fake        0         0    0      0       0          0           0   

   fulfilling  ...  driven  arabia  adios  arbitrary  cake  drills  handler  \
0           0  ...       0       0      0          0     0       0        0   
1           0  ...       0       0      0          0     0       0        0   
2           0  ...       0       0      0          0     0       0        0   
3           0  ...       0       0      0          0     0       0        0   
4           0  ...       0       0      0          0     0       0        0   

   hills  donation  toilet  
0      0         0       0  
1      0         0       0  
2      0         0       0  
3      0         0       0  
4      0         0       0  

[5 rows x 7718 columns]

## Calculating constants first

Now i need to find the probability of fake news given title and real news given title. So we need to find:
* p(fake)
* p(real)
* Nfake
* Nreal
* Nvocabulary

In [15]:
fake_title = train_clean[train_clean['label'] == 'Fake']
real_title = train_clean[train_clean['label'] == 'Real']

In [16]:
p_fake = len(fake_title)/len(train_clean)
p_real = len(real_title)/len(train_clean)

print('p(fake):', p_fake)
print('p(real):', p_real)

p(fake): 0.5044543429844098
p(real): 0.4955456570155902


In [17]:
n_words_per_fake_title = fake_title['title'].apply(len)
n_fake = n_words_per_fake_title.sum()

n_words_per_real_title = real_title['title'].apply(len)
n_real = n_words_per_real_title.sum()

n_vocabulary = len(vocabulary)

alpha = 1

print('Nfake:', n_fake)
print('Nreal:', n_real)
print('Nvocabulary:', n_vocabulary)

Nfake: 28392
Nreal: 18753
Nvocabulary: 7716


## Calculating parameters
Now that we have the constant terms calculated above, we can move on with calculating the parameters P(wi|Fake) and P(wi|Real). Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

In [18]:
parameters_fake = {unique_word: 0 for unique_word in vocabulary}
parameters_real = {unique_word: 0 for unique_word in vocabulary}

for word in vocabulary:
    n_word_fake = fake_title[word].sum()
    p_word_fake = (n_word_fake + alpha) / (n_fake + alpha * n_vocabulary)
    parameters_fake[word] = p_word_fake
    
    n_word_real = real_title[word].sum()
    p_word_real = (n_word_real + alpha) / (n_real + alpha * n_vocabulary)
    parameters_real[word] = p_word_real

## Classifying a new headline

In [19]:
def classify(headline):
    
    headline = re.sub('\W', ' ', headline)
    headline = headline.lower()
    headline = headline.split()
    
    p_fake_given_headline = p_fake
    p_real_given_headline = p_real
    
    for word in headline:
        if word in parameters_fake:
            p_fake_given_headline *= parameters_fake[word]
        
        if word in parameters_real:
            p_real_given_headline *= parameters_real[word]
            
    print('P(fake|headline):', p_fake_given_headline)
    print('P(real|headline):', p_real_given_headline)
    
    if p_fake_given_headline > p_real_given_headline:
        print('Label: Fake')
    elif p_fake_given_headline < p_real_given_headline:
        print('Label: Real')
    else:
        print('Equal probabilities, have a human classify this')

In [20]:
classify('Obama issues call to action in eulogy for Lewis, links Trump to foes of civil rights in 1960s')

P(fake|headline): 9.063723844475128e-52
P(real|headline): 8.25304087114931e-51
Label: Real


In [21]:
classify('Biden Campaign Whittles VP Shortlist Down To Either Woman Or Man With Long Hair')

P(fake|headline): 2.9246202514610007e-40
P(real|headline): 1.898196617709361e-41
Label: Fake


## Classifying test samples

In [22]:
def classify_test(headline):
    
    headline = re.sub('\W', ' ', headline)
    headline = headline.lower()
    headline = headline.split()
    
    p_fake_given_headline = p_fake
    p_real_given_headline = p_real
    
    for word in headline:
        if word in parameters_fake:
            p_fake_given_headline *= parameters_fake[word]
        
        if word in parameters_real:
            p_real_given_headline *= parameters_real[word]
    
    if p_fake_given_headline > p_real_given_headline:
        return 'Fake'
    elif p_fake_given_headline < p_real_given_headline:
        return 'Real'
    else:
        return 'Needs human classification'

In [23]:
test['predicted'] = test['title'].apply(classify_test)
test.head()

title  \
0                Saudi-led coalition allows first aid ship into Yemen's Hodeidah port: local officials   
1   Trump Administration In FREE FALL After ‘Concrete Evidence’ Linking Trump To Russia Emerges (DE...   
2        Watch Robert Reich’s Brain Nearly MELT After Republican Says CIA Is ‘Attacking’ Trump (VIDEO)   
3                                                U.S. Navy may raise current 308-ship target for fleet   
4          BOOM! DR ALVEDA KING Scolds Sen. Liz Warren: We won’t accept racist bait and switch [Video]   

  label predicted  
0  Real      Real  
1  Fake      Fake  
2  Fake      Fake  
3  Real      Real  
4  Fake      Fake

In [24]:
print('Test labels:')
print(test['label'].value_counts())
print('-' * 20)
print('Predicted labels:')
print(test['predicted'].value_counts())

Test labels:
Fake    490
Real    408
Name: label, dtype: int64
--------------------
Predicted labels:
Fake    494
Real    404
Name: predicted, dtype: int64


The amount of real news decreased between the test labels and the predicted labels. While the amount of fake news increased. There wasn't any headline that required a human classification

In [25]:
test[(test['label'] == 'Real') & (test['predicted'] == 'Fake')]

title  \
15                Republican Romney to make 'major speech' on 2016 presidential race: Fox   
35                           For Republican Rubio, a moment of truth in race to lead U.S.   
73                                     NRA calls for more regulation after Vegas shooting   
74                       Trump says he is 'very, very close' to making Fed chair decision   
93                    Syrian militant group releases video of leader apparently uninjured   
108                       Martin Luther King's daughter says 'God can triumph over Trump'   
112                             Rand Paul's accused attacker pleads not guilty to assault   
134                          Trump accuses Cruz of stealing Iowa caucuses through 'fraud'   
141                                   Obama, Clinton scold Trump over proposed Muslim ban   
158                                          Factbox: Why the Zika virus is causing alarm   
180              Trump's about-face on fund-raising seems unlikely to dent his popularity   
186                                  Jeb Bush endorses Ted Cruz for Republican nomination   
214                              New poll shows Philippine president still hugely popular   
220                       Indonesian police detain 22 over violent anti-communist protest   
228                          Stop fighting over Brexit and get real, Jim O'Neill tells UK   
256                          New York City mayor tells Trump about fear over his policies   
283  Seattle judge grants nationwide temporary restraining order on Trump immigration ban   
337        Japanese girl says school forced her to dye hair black, sues government: media   
378                       Many 'lost' voters say they have found their candidate in Trump   
379                             Russia hunting four people behind huge bomb hoax campaign   
383                   Russia's response to Trump leak reports: don't read U.S. newspapers   
413                    'A better future' - Britain's May tries to rally her Conservatives   
459                       McCain to Trump on Russian hacking: 'The facts are there' - CBS   
476                                  Trump hustled off stage in Nevada by security agents   
518                              Trump adds Washington lawyer John Dowd to his legal team   
616                              Mexico City gets unsteadily back on its feet after quake   
651                                             A low-key Trump ends social media silence   
665                                           Democrats chide Sanders after Nevada mutiny   
682                  Trump says son is 'innocent' over emails about Russian campaign help   
695                          'Joe the Plumber' praises Trump, cites his 'beautiful women'   
709                 Putin's dinner with Michael Flynn: 'I didn't even really talk to him'   
716                     Cruz gets Jeb Bush backing as Republican elites try to stop Trump   
779                                          Baby banned from Japanese municipal assembly   
787                                Liberal groups urge non-violent movement to stop Trump   
801                                            Clinton's pantsuits inspire flashmob video   
860                       Obama on growing friendship with Trudeau: 'What's not to like?'   
868                      A poet in the dock as Israel hunts 'lone wolf' assailants online   
879                         Florida legislature poised to bolster 'Stand Your Ground' law   

    label predicted  
15   Real      Fake  
35   Real      Fake  
73   Real      Fake  
74   Real      Fake  
93   Real      Fake  
108  Real      Fake  
112  Real      Fake  
134  Real      Fake  
141  Real      Fake  
158  Real      Fake  
180  Real      Fake  
186  Real      Fake  
214  Real      Fake  
220  Real      Fake  
228  Real      Fake  
256  Real      Fake  
283  Real      Fake  
337  Real      Fake  
378  Real      Fake  
379  Real      F

There are 21 headlines that apparently were misclassified. I searched for them on google and found that all of them are headlines on the Reuters website and other sites.

In [26]:
test[(test['label'] == 'Fake') & (test['predicted'] == 'Real')]

title  \
10            Clinton’s ‘No-Fly Zone’ over Syria Will Not “Save Lives” – It Will Lead to War with Russia   
27                                     Sanders Campaign Fights Back After Ohio Bans Youth From Primaries   
39   #CrookedHillary’s Karma! NEW POLL Shows Trump Takes Lead In BLUE STATES: Michigan, Wisconsin, Ne...   
49                                      NY Lawmakers Compare MMA Fighting To ‘Gay P*rn’ And Police Abuse   
57                       House GOP Holds Zika Bill Hostage, Proposes Plan To Slash Birth Control Funding   
80                           Trump Slapped With Lawsuit For Refusing To Release White House Visitor Logs   
83                                    Trump Considers ‘Terminating NAFTA’ With Executive Order (DETAILS)   
111  BOOM! Fed Judge Ruling UNBLOCKS Trump Travel Ban…Asks ACLU Lawyer: “Where Does It Say Muslim Cou...   
115                  Trump Owes A Foreign Bank At Least $100 Million – And It’s Fighting U.S. Regulators   
149                     Before Scalia’s Body Is Even Cold, Republicans Vow To Obstruct ANY Obama Nominee   
164                                           Trump Conducts Sting Operation on US Intelligence Services   
173                    California Assembly Passes MAJOR Legislation To Keep Trump Off The Ballot In 2020   
234                                FEMA DENIES AID TO BALTIMORE FOR RIOT DAMAGES – MAYOR PLANS TO APPEAL   
242      “SENIOR U.S. OFFICIAL” Releases Damaging Information On Russian Involvement In Syria Gas Attack   
274                                  WOW! CALIFORNIA SUPREME COURT Case Decides ‘Right to Vote On Taxes’   
286  ILLEGAL ALIEN CRIMINALS ON HUNGER STRIKE IN AZ PRISON Give List Of Demands Including Removal Of ...   
298                   BREAKING NEWS: Vladimir Putin Retaliates After New Sanctions Are Slapped On Russia   
347  NUCLEAR SHOWDOWN? NAVY SEAL TEAM That Killed Osama Bin Laden To Take Part In Military Drills Aga...   
355                   Kim Jong Un And North Korean Media Enthusiastically Want Trump To Become President   
375                Former Campaign Finance Co-Chair BLASTS Christie For Trump Endorsement, Pulls Support   
450                          WATCH: 3 PRESIDENTS Before Trump Promised To Move U.S. Embassy To Jerusalem   
464                                                         The Genealogy of Trump’s U-Turn on Palestine   
513                          WATCH: 3 PRESIDENTS Before Trump Promised To Move U.S. Embassy To Jerusalem   
526       Half The World’s Salmon Supply Threatened As Trump’s EPA Withdraws Alaskan Mining Restrictions   
570                              ‘#1 In Bigotry’: Twitter EVISCERATES Mississippi Gov. Over Anti-Gay Law   
613                             DEMOCRATS FUMING Over Vote To Keep “Hurtful” Word In Library Of Congress   
621                                China Fires Back As Trump Drama ESCALATES Before He Even Takes Office   
637  EU LEADERS PLEDGE EXTRA €1 Billion In Aid To Refugees…Slovakia Will Take EU To Court Over Forced...   
647                 BOMB THREAT AND CRITICISM HITS HISPANIC BBQ OWNER’S PLANS FOR WHITE APPRECIATION DAY   
649                             Tom Cotton Accepted $1 Million From Israel To Sabotage Iran Nuclear Deal   
753  HYPOCRISY: DEMOCRAT SENATOR Criticizes President Trump For Moving U.S. Embassy To Jerusalem Afte...   
784                       DRUNK REFUGEES Spit On And Bite German Nurses: Force Hospital To Hire Security   
809                                      NANNY STATE GOVERNMENT S’MORES: NO CHOCOLATE AND NO MARSHMALLOW   
832           Obama Finally Takes Action: Kicks 35 Russian Diplomats Out Of U.S. Over Election Tampering   

    label predicted  
10   Fake      Real  
27   Fake      Real  
39   Fake      Real  
49   Fake      Real  
57   Fake      Real  
80   Fake      Real  
83   Fake      Real  
111  Fake      Real  
115  Fake      Real  
149  Fake      Real  
164  Fake      Real  
173  Fake      Real  
234  Fake      Real  
242  Fake 

On the other side, 9 fake news were reported as real news. A search on google showed that the majority of them don't appear at all. And at least 3 of them were click baits or were misclassified. For example the headline "Report: Facebook’s Zuckerberg Gave FBI’s Mueller Info for Russia Investigation" i found multiple articles in websites like cnbc, wsj (Wallstreet journal), politico.com and vanityfair that had the same headline or similar. Maybe the content of the article is the one that's giving fake news

Now let's meassure the accuracy of the Fake news classifier

In [27]:
correct = 0
total = test.shape[0]

for row in test.iterrows():
    row = row[1]
    if row['label'] == row['predicted']:
        correct += 1
        
print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', correct / total)

Correct: 826
Incorrect: 72
Accuracy: 0.9198218262806236


The algorithm has a 93% of accuracy, wich is good. However, it looks like analyzing only the title isn't enough to do a proper classification of fake/real news, there's a chance that using a bigger portion of the full dataset, the accuracy might go down. A text analysis is required, but to be able to do it with this algorithm a higher processing power is required. Also there's a need to apply things like stemming and feature selection to have a better accuracy.